toolkit/tools/internal/pkggraph/pkggraph.go

```diff
@@ -1434,3 +1574,64 @@ func findAllRPMS(rpmsToFind []string) (foundAllRpms bool, missingRpms []string)
 
 	return
 }
+
+// findAllRPMSUpstream returns true if all RPMs requested are found in PMC/repo.
+//  This is used to check if all RPMs are available in PMC/repo before starting a build.
+//  needs repo-list, manifest files rpmrepocloner
+func findAllRPMSUpstream(rpmsToFind []*PkgNode, outDir, tmpDir, workerTar, existingRpmsDir, existingToolchainRpmsDir string, usePreviewRepo, ignoreVersionToResolveSelfDep bool, repoFiles []string) (foundAllRpms bool) {
+
+        // Create the worker environment
+// Initialize initializes rpmrepocloner, enabling Clone() to be called.
+//   - destinationDir is the directory to save RPMs
+//   - tmpDir is the directory to create a chroot
+//   - workerTar is the path to the worker tar used to seed the chroot
+//   - existingRpmsDir is the directory with prebuilt RPMs
+//   - prebuiltRpmsDir is the directory with toolchain RPMs
+//   - usePreviewRepo if set, the upstream preview repository will be used.
+//   - repoDefinitions is a list of repo files to use when cloning RPMs
+
+        cloner := rpmrepocloner.New()
+	err := cloner.Initialize(outDir, tmpDir, workerTar, existingRpmsDir, existingToolchainRpmsDir, usePreviewRepo, repoFiles)
+        if err != nil {
+                logger.Log.Errorf("Failed to initialize RPM repo cloner. Error: %s", err)
+                return
+        }
+        defer cloner.Close()
+
+        err = cloner.AddNetworkFiles("", "")
+        if err != nil {
+                logger.Log.Panicf("Failed to customize RPM repo cloner. Error: %s", err)
+        }
+
+	/*assuming allRPMs found, to be set to false on error return or when a package is not found*/
+	foundAllRpms = true
+	//for each rpmToFind in rpmsToFind, do
+	for _, node := range rpmsToFind {
+               logger.Log.Debugf("Searching for a package which supplies: %s", node.VersionedPkg.Name)
+               // Resolve nodes to exact package names so they can be referenced in the graph.
+		if ignoreVersionToResolveSelfDep {
+			/*clear Version information so that any available version and release can be used*/
+			clearVersion(node.VersionedPkg)
```

Comment:
>clearVersion(node.VersionedPkg) [](http://example.com/codeflow?start=3&length=31)

We're using node pointers here, so we're now changing the information in the original graph.

And a more important question: why do we want to ignore version limitations for dependencies? Building with a wrong version of the dependency may cause build failures or produce invalid output. The version restrictions from the spec should not be usually ignored.

Reply:
We can choose to ignore version limitations when the package requires itself to build (so that an older version of the package can be used to build the new version). This is disabled by default and can be enabled when required using IGNORE_VERSION_TO_RESOLVE_SELFDEP flag.